atlp-rwanda/code-ninjas-be
# #181414761 feature(user-profile):create user user profile

### What does this PR do?
Create user profile after registration.

### Description of Task to be completed?
- Create user profile route
- Tested the route
- Documented the route

### How should this be manually tested?
To test the changes manually:
* clone repo on your local machine
* switch to develop branch  by typing the command ```git checkout ft-user-profile-181414761```
* Run: ```yarn install```: to install all dependencies
* Create .env file
* Create a ```.env``` and the variables as mentioned in the ```.env.example``` file.
* Open the host address using an HTTP client such as ``postman``` or any other.
* Make ``GET`` or  ``PUT`` on  ``/api/v2/user/profile/:id`` to retrieve user's profile and/or update it.

### What to expect after running the commands:
* _On successful, A user will get a status of ``200`` for profile fetch and delete and ``201`` for profile update_

* _If the user doesn't exist, you'll a message that says: ``No such User was found`` with the status of: ``404``._

### Any background context you want to provide?
- N/A

``` Useful links ```:
- N/A

### What are the relevant pivotal tracker stories?
* https://www.pivotaltracker.com/story/show/181414761

### Screenshots (if appropriate)
- N/A

### Questions:
- N/A


---


### Task implementation plan

#1. Name of task
- Profile Page Settings

#2. Pivot Tracker ID

- [181414761](https://www.pivotaltracker.com/story/show/181414761)

#3. Branch Name

- ft-create-user-profile-181414761

#4. Description of task (Paraphrase what we have in Pivot Tracker)

* After signup on Barefoot Nomad, a user should be able to have their profile created and may update the account to add some more information.


#5. Acceptance criteria (Checks to be made)

- [ User should be able to Edit my profile, but should NOT be able to edit another user's profile]

```Scenario```: Given User registers to be a member of Barefoot Nomad.Registration is done.User Profile is created.A timestamp is associated to the profile

#6. Database schema (Migration & models)
 
<img width="431" alt="Screen Shot 2022-04-18 at 18 40 37" src="https://user-images.githubusercontent.com/75941936/163841443-2b591819-30e9-4b7d-af84-e41fa033f7e0.png">

#7. Validations point (Input to validate)
* Will use Joi validator to check:

```For profile```:

- Gender - string, valid(‘Male’, ’Female’, ’Prefer Not say’)
- Date of Birth - date, iso, 
- Gender: string,
- Age: number,
- Nationality: string,
- Preferred Language: string
- Preferred Currency: string
- Department: stringAddress: string
- Line Manager: string
- Phone Number: number
- Marital Status: string, valid(‘Single’, ‘Married’, ‘Widowed’)

#8. Logic on the access (Explaining who is allowed to CRUD (if applicable) the entities you are working on)

* User: Allowed to view and update their profile after registration

#9. Logic on actual Implementation

- Update User model for user 
- Setup the validations using Joi
- Use middleware to check if user is logged in, is profile owner
- Create controllers for profile retrieve and update
- Create the route for profile retrieve and update
- Test the routes with mocha & chai
- Document the endpoint with swagger

#10. Endpoints Format: {url}/api/user/….

GET   - /profile/:id: Get single profile - Owner
PUT   - /profile/:id: update user data - Owner
